Implementation of class to represent a parametric region in space

friyaz opened the pull request:
Added an implementation of ParamRegion class.
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
#19320 

#### Brief description of what is fixed or changed
An object of the ParametricRegion class should represent a parametric region in space. This object can then be used to perform scalar/vector integration over parametric regions.

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* vector
  * Added class to represent a parametric region in space.
<!-- END RELEASE NOTES -->

friyaz committed:
Added a rough implementation of ParamtericRegion in sympy.vector

The class ParametricRegion represents a parametric region in space.
Objects of this class will used to perform vector/scalar integration.

friyaz committed:
remove trailing whitespace from __init__.py

friyaz committed:
add pi to docstring

friyaz committed:
Added tests in test_args.py

Upabjojr commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,48 @@
+        >>> from sympy.vector import CoordSys3D, ParametricRegion
+        >>> r, theta = symbols("r theta")
+        >>> C = CoordSys3D('C')
+        >>> circle = ParametricRegion(C, r*cos(theta), r*sin(theta), (theta, 0, 2*pi))
```
Is the parametric region defined within a `CoordSys3D` object? Can this be made optional? It's easier to use if you don't have to define too many objects.

friyaz replied:
Yes, It will be easier if a user does not have to pass the `CoordSys3D` object. But we need to determine base scalars otherwise the definition of a parametric region is incomplete. A parametric representation is just defining base scalars in terms of some parameters. 

We can avoid passing the `CoordSys3D` object if we use the base scalars of the vector/scalar field  which needs to be integrated. So, if a user does not pass a `CoordSys3D` object, then SymPy will use the base scalars of the vector/scalar field used for integration. 

```python
C = CoordSys3D('C')
R = C.locate_new('R', 3*C.i + 4*C.j + 5*C.k)
p = ParametricRegion(t, t**2, (t, 0, 1))
ParametricIntegral(C.x*C.y*C.z*C.i, p) # Define p in terms of C.x, C.y and C.z 
ParametricIntegral(R.x*R.y, p) # Define p in terms of R.x, R.y and R.z  
```

Upabjojr replied:
What about using a convention similar to Lambda?

friyaz replied:
@Upabjojr Can you please give an example.

Upabjojr commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,48 @@
+        self._definition = list()
+        self._bounds = list()
+        self._parameters = list()
+        self.system = system
```
`self._system = system` would be the standard way... then
 ```
@property 
def system(self):
    return self._system
```

This make sure that `system` cannot be modified by the user.

friyaz replied:
Yes, I will correct it.

Upabjojr commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,48 @@
+                self.definition.append(arg)
+
+        if len(self.definition) > 3:
+            raise ValueError("Only 3-D regions are supported")
```
Why this restriction?

friyaz replied:
I think the vector module supports only 3-D Coordinate frame. If we cannot define a vector field of more dimensions, then why allow defining a higher dimension.

Upabjojr replied:
I don't see the point of having this restriction. ParametricRegion shouldn't only be defined in 3D.

friyaz replied:
OK, I have removed these conditions in the new commit.

oscarbenjamin commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,48 @@
+    """
+    Represents a parametric region in space.
+    """
+    def __init__(self, system, *args):
```
Basic subclasses should use `__new__` rather than `__init__` and should access their data from `self.args`.

friyaz replied:
Thanks, I noticed this. All subclasses of `Basic` in sympy use `__new__`.

What is the reasoning behind this? AFAIK, we should use `__new__` if we intend to return objects of different types based on some conditions.

jksuom replied:
> return objects of different types based on some conditions.

This is what is often done in SymPy. For example, `Rational.__new__` may return an instance of `Integer` and `Integer.__new__` may return `S.One`.

friyaz commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,48 @@
+
+    @property
+    def bounds(self):
+        return self._bounds
```
Maybe we should return bounds in the form of dict().
```
{t: (0, 1)}
```
This form will be more useful for 'ParametricIntegral` class.

Upabjojr replied:
Pay attention that dictionaries are not ordered. Maybe we care about the order of parameters?

friyaz replied:
OK. It is better to use a list or tuple.

friyaz committed:
Use __new__ instead of __init__

Subclasses of basic should use __new__ instead of __init__
Also, removed check for length of defintion

friyaz committed:
Raise ValueError when parameters specified incorrectly.

friyaz committed:
Add test cases for ParametricRegion class

friyaz committed:
Use self.args

Upabjojr commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,51 @@
+    >>> r, theta = symbols("r theta")
+    >>> circle = ParametricRegion(r*cos(theta), r*sin(theta), (theta, 0, 2*pi))
+    """
+    def __new__(cls, *args, system=None):
```
What about defining some precise signature? `*args` is a bit ambiguous.

friyaz replied:
Yes, `*arg` is ambiguous. I used so we can define base scalars and bounds without nesting them in a tuple.

Upabjojr replied:
what about `def __new__(cls, parameters_or_coordsys, mapping_tuple, **kwargs):` ?

friyaz replied:
```python
ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), r=(0,1), theta=(0, pi))
ParametricRegion(C, (C.y, 3, -3), C.y=(4,6))
```
I don't think the User should pass a tuple of parameters or CoordSys3d object. 
We can easily determine the parameters using key values of the dictionary.

But we should also add an optional parameter for specifying the base scalars(see discussion above on `CoordSys3D` object) if a user does not want to user base scalars of the scalar/vector field. You mentioned using some lambda convention. Please explain that part.

What about this API @Upabjojr ?
```python
def __new__(cls, system=None, mapping_tuple, **kwargs): 
```
Example
```python
ParametricRegion(((r*cos(theta), r*sin(theta)), r=(0,1), theta=(0, pi))
ParametricRegion(system=C, (C.y, 3, -3), C.y=(4,6)) 
```

Upabjojr replied:
`r=(0, 1), theta=(0, pi)` is not very intuitive, what about a dict of bounds?

What about this:
```python
ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), limits={r: (0,1), theta: (0, pi)})
ParametricRegion(C, (C.y, 3, -3), limits={C.y: (4,6)}) 
```

> You mentioned using some lambda convention. Please explain that part.

Define variable tuple, then mapping tuple.

friyaz replied:
I am still not sure why a user needs to pass a tuple specifying parameters. SymPy can determine them using the dict of bounds. Will it be more intuitive for users?
```python
ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), limits={r: (0,1), theta: (0, pi)})
```

Upabjojr replied:
Well, a dict does not have an order, while a tuple does. Plus, it's the same definition of lambdas:
```python
lambda r, theta: (r*cos(theta), r*sin(theta))
```

friyaz replied:
+1 for passing parameter_tuple then.

friyaz replied:
@Upabjojr What are your thoughts on the system parameter?
As discussed above, We have two options:
* Assume that the mapping of param region is to base vectors of the scalar/vector field.
```python
p = ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), limits={r: (0,1), theta: (0, pi)})
Integral(C.x**2*C.y, p)
# C.x = r*cos(theta),  C.y = r*sin(theta), C.z = C.z
Integral(R.x*R.y*R.z, p)
# R.x = r*cos(theta),  R.y = r*sin(theta), R.z = R.z
```
* The other option is to allow users if they wish to specify the coordinate system.
```python
p = ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), limits={r: (0,1), theta: (0, pi)}, system = C)
Integral(R.x*R.y*R.z, p)
# C.x = r*cos(theta),  C.y = r*sin(theta), C.z = C.z
```

Upabjojr commented on sympy/vector/tests/test_parametricregion.py:
```diff
@@ -0,0 +1,50 @@
+
+def test_parametricregion():
+
+    point = ParametricRegion(3, 4)
```
What about `ParametricRegion((), (3, 4))` to define a point? That is... first argument is the tuple of parameters (in this case an empty tuple), second one the mapping.

I think this would be more intuitive. What do you think?

friyaz replied:
@Upabjojr How will we get the definition of base scalars in terms of parameters? Also,  if we fix the number of arguments to 2, we have to use nested tuples to define the bounds for more than 2 parameters
```python
ParametricRegion((r, theta), ((r, 0, 2), (theta, 0, pi))
```

friyaz replied:
This one looks good to me. It is still not very simple but we have to keep some minimum parameters. They are essential for defining a parametric region.

Upabjojr replied:
It's the same concept for integrals:
```python
Integral(x, x)  # ==> indefinite
Integral(x, (x, 0, 10))  # ==> definite
```

If you have a better API idea, you're welcome to suggest.

Upabjojr commented on sympy/vector/tests/test_parametricregion.py:
```diff
@@ -0,0 +1,50 @@
+    assert point.bounds == []
+    assert point.system == None
+
+    l = ParametricRegion(C.y, (C.y, -3, 3), system=C)
```
Maybe we could have either `ParametricRegion((y,), (y, -3, 3))` or `ParametricRegion(C, (C.y, -3, 3))`.

That is, if the first argument is a tuple, then it's assumed to be a list of variable-parameters. If it is a `CoordSys3D` object, then use its scalars as parameters.

This means that `ParametricRegion` would always take 2 arguments when called. What do you think of this API?

friyaz replied:
> That is, if the first argument is a tuple, then it's assumed to be a list of variable-parameters. If it is a CoordSys3D object, then use its scalars as parameters.

Yes, this API looks simple and easy but
```python
ParametricRegion(C, (C.y, -3, 3))
```
I am not able to understand what this region represents. 
 For me, `ParametricRegion(C.y, (C.y, -3, 3), system=C)` represents the line y=x where y varies from -3 to 3. Does this represent the same region?

Upabjojr replied:
> For me, `ParametricRegion(C.y, (C.y, -3, 3), system=C)` represents the line y=x where y varies from -3 to 3. Does this represent the same region?

Oh, I see. With `ParametricRegion(C, (C.y, -3, 3))` I think of the line y = -3 and z = 3, with `C.y` being the parameter varying along the x-axis (x-axis because it's in the first position).

friyaz replied:
> line y = -3 and z = 3, with C.y being the parameter varying along the x-axis

I think we can define this region without using the coordinate scalar. How is this different from 
```python
ParametricRegion((t), (t, -3, 3))
```

Upabjojr replied:
`(t)` is not a tuple, `(t,)` is. Unfortunately Python has some strange syntax.

friyaz replied:
Okay. Then, to make it easier for users, we can put a condition like
```python
if not isinstance(parameter, tuple):
    parameter = tuple(parameter)
```

Upabjojr commented:
Also have a look at the image `ImageSet` class. It's vaguely similar to this one, maybe some of its code could be reused?

friyaz committed:
Modified API of ParametricRegion class

Upabjojr commented on sympy/vector/parametricregion.py:
```diff
@@ -0,0 +1,47 @@
+    >>> r, theta = symbols("r theta")
+    >>> circle = ParametricRegion((r, theta), r*cos(theta), r*sin(theta), limits={theta: (0, pi)})
+    """
+    def __new__(cls, parameters, definition, limits, system=None):
```
What about setting `limits` to be optional? For example, with a default `limits=None` argument.

friyaz replied:
Yes, It will be better. I will do it.

friyaz committed:
Made limits optional

Upabjojr commented on sympy/core/tests/test_args.py:
```diff
@@ -4918,14 +4918,19 @@ def test_sympy__vector__orienters__QuaternionOrienter():
     C = CoordSys3D('C')
     assert _test_args(BaseScalar(0, C, ' ', ' '))
 
-
```
We usually leave two empty lines on the base scope...

Upabjojr commented:
Tests keep falling.

Upabjojr commented:
If you fix the tests, I will merge this.

friyaz commented:
OK, I will fix the test and improve the docstring.

friyaz committed:
Modified API of ParametricRegion class, improved docstring and Fixed failing tests of core/test/test_args

sylee957 commented on sympy/vector/tests/test_parametricregion.py:
```diff
@@ -0,0 +1,56 @@
+    assert ellipse.limits == {t: (0, 8)}
+
+    cylinder = ParametricRegion((r, theta, C.z), (cos(theta), r*sin(theta), C.z), {theta: (0, 2*pi), C.z: (0, 4)})
+    assert cylinder.parameters == (r, theta, C.z)
```
I see you have introduced `parameters_or_coordsys` only because you have to deal with coordinate systems `C`, but as you mix coordinate scalars and ordinary symbols in this way, you can always specify the region regardless of the ambient space of the coordinate systems and it loses no generality to use ordinary symbols only (`x, y, z` instead of `C.x, C.y, C.z`).
And also what would be the necessary reason to pass `C` in place of `(C.x, C.y, C.z)`

Upabjojr replied:
> And also what would be the necessary reason to pass `C` in place of `(C.x, C.y, C.z)` if they are implicitly converted any way?

Shortcut?

friyaz replied:
> it loses no generality to use ordinary symbols like x, y, z everywhere in place of coordinate systems.

Yes, I think we can replace base scalars with any variable. Suppose we want to represent a rectangle:
```python
ParametricRegion(C, (C.x, C.y), {C.x: (1, 2), C.y: (3, 5)})
```
We can write this as
```python
ParametricRegion((x, y), (x, y), {x: (1, 2), y: (3, 5)})
# C.x = x,  C.y = y, C.z = C.zd
```

friyaz replied:
Just to make this explicit,
When we define the definition tuple ..suppose `(r*cos(theta), r*(sin(theta), r*sin(phi))`, this implies
```
C.x = r*cos(theta), C.y = r*(sin(theta), C.y = r*sin(phi)
```
These base scalars are assumed. They may be determined from the base scalars of the vector/scalar field for integration.
The only reason I think we may need a coordinate system variable other than for shortcut is when one needs to specify base scalars (instead of using base scalars of the vector/scalar field).  

```
p = ParametricRegion((r, theta), (r*cos(theta), r*sin(theta)), limits={r: (0,1), theta: (0, pi)})
Integral(C.x**2*C.y, p)
# C.x = r*cos(theta),  C.y = r*sin(theta), C.z = C.z
Integral(R.x*R.y*R.z, p)
# R.x = r*cos(theta),  R.y = r*sin(theta), R.z = R.z
```

sylee957 replied:
What happens if it is `Integral(R.x*C.y, p)`?
Perhaps, `p` can have its own base scalars?

friyaz replied:
Yes, I am in favor of this. But there should be some relationship between the coordinate system of vector/scalar field and the coordinate system of the parametric region. For calculating the integral, we need to replace base scalars in the field with parameters. This requires some connection between them.

@Upabjojr What do you think?

Upabjojr replied:
My original thought was just to use parameters. Mixing parameters and base scalars in the tests is a bit strange, but it's OK with me.

friyaz replied:
I think for the tests, we can leave them as it is. But for users, it should be preferred to not mix them.

friyaz committed:
Remove test case with mixed base scalars and parameters
